File: sync/src/main/scala/com/fan/mysql/parser/GtidEventParser.java
package com.fan.mysql.parser;


import com.fan.mysql.binlog.BinlogEventParser;
import com.fan.mysql.dbsync.BinlogContext;
import com.fan.mysql.dbsync.LogBuffer;
import com.fan.mysql.event.BinlogEvent;
import com.fan.mysql.event.EventHeader;
import com.fan.mysql.event.impl.GtidEvent;

public class GtidEventParser implements BinlogEventParser {

    public BinlogEvent parse(LogBuffer buffer, EventHeader header, BinlogContext context) {
        GtidEvent event = new GtidEvent(header);
        buffer.getInt8(); // commit flag, always true
        event.setSourceId(buffer.getData(16));
        event.setTransactionId(buffer.getLong64());
        if (buffer.hasRemaining()) {
            event.setLogicalTimestamp(buffer.getInt8());
            event.setLastCommitted(buffer.getLong64());
            event.setSequenceNumber(buffer.getLong64());
        }
        return event;
    }

}

File: async/src/main/scala/com/fan/mysql/async/binlog/parse/GtidEventParser.scala
package com.fan.mysql.async.binlog.parse

import com.fan.mysql.async.binlog.event.impl.GtidLogEvent
import com.fan.mysql.async.binlog.event.{BinlogEvent, EventHeader}
import com.fan.mysql.async.binlog.{BinlogDumpContext, BinlogEventParser}
import io.netty.buffer.ByteBuf

class GtidEventParser extends BinlogEventParser {
  override def parse(buffer: ByteBuf,
                     header: EventHeader,
                     context: BinlogDumpContext): BinlogEvent = {
    val event = new GtidLogEvent(header)

    buffer.readByte() // commit flag, always true

    event.setSourceId(buffer.readBytes(16).array())
    event.setTransactionId(buffer.readLong())

    if (buffer.isReadable) {
      event.setLogicalTimestamp(buffer.readByte())
      event.setLastCommitted(buffer.readLong())
      event.setSequenceNumber(buffer.readLong())
    }

    event
  }
}
